--- libguestfs-1.2.7.orig/fish/fish.c	2010-07-07 18:48:21.905925442 +0100
+++ libguestfs-1.2.7/fish/fish.c	2010-07-07 18:49:25.402675361 +0100
@@ -403,6 +403,17 @@
 
   /* If we've got mountpoints, we must launch the guest and mount them. */
   if (mps != NULL) {
+    /* RHBZ#612178: If --listen flag is given, then we will fork into
+     * the background in rc_listen().  However you can't do this while
+     * holding a libguestfs handle open because the recovery process
+     * will think the main program has died and kill qemu.  Therefore
+     * don't use the recovery process for this case.  (A better
+     * solution would be to call launch () etc after the fork, but
+     * that greatly complicates the code here).
+     */
+    if (remote_control_listen)
+      guestfs_set_recovery_proc (g, 0);
+
     if (launch (g) == -1) exit (EXIT_FAILURE);
     mount_mps (mps);
   }
